### ISSUE TITLE
Serialize the unsaved changes of files not associated with a worktrees

### ISSUE THREAD

**Issue by JosephTLyons** [zed-industries/zed#15098]
### Check for existing issues

- [X] Completed

### Check for existing issues

- [X] Completed

### Describe the feature

Since the following PRs, Zed now restores unsaved content:

- https://github.com/zed-industries/zed/pull/13546
- https://github.com/zed-industries/zed/pull/14965

There is a limitation currently. Zed will serialize the changes for files if there are worktrees / projects in the project panel, but it will not do so for instances of Zed that have no worktrees / projects in the project panel - supporting this case would be a nice addition to the workspace serialization story.

**Comment by selfish** [comment]
Very interested in this.
Thanks!!

**Comment by emadpres** [comment]
This issue is literally the only reason I'm still using _Sublime_ as my daily driver. I'd love to draft some notes or paste a code snippet, and still be able to quit the app without being bugged for saving.

**Comment by emadpres** [comment]
Any update on this?

**Comment by fboleto** [comment]
A lot of us are still using VS code, or sublime due to the lack of this feature. Anyone has a timeline for this feature to be released?

**Comment by AldoMX** [comment]
The dev team has explained that they architected the editor around the projects feature, so this request is not feasible without a large refactor that decouples the projects feature.

They hadn't prioritized this request due to the effort required by that refactor.

The workaround is to create an empty folder and open it as a project.

If you ask me, I would suggest to the dev team to use `/dev/null` as the default project, and leave the refactor for later, people need this feature today, not until 2030.

**Comment by willik** [comment]
Just got bit by this and lost several files.. bye bye zed.

**Comment by marksweiss** [comment]
Another vote for this feature, and an observation. Which is, is Zed competing with IDEs or text editors? Because centering this functionality around having to have a file in a project is clearly putting Zed philosophically on the IDE side. But ... almost all the rest of what it does and how is like a modern take on _building up_ _if you want to_ the core features of an IDE In a lighter-weight editor, as you would in Vim. For me, the exact position Zed should take is as a complement to IDEs, not a competitor. For when you have ephemeral text editing tasks, light-weight coding tasks that aren't yet in a repo, and code-reading tasks that are outside your main repo/IDE or you just want to do things more quickly. But if you don't support this whole set of uses, people can't ADD Zed to their workflow NEXT TO their IDE. I'm going to do the workaround of an empty project, because I have to, but it's silly and points to the problem with the initial design decision. My suggestion is to support these use cases first class, because they are "what's left" outside of IDEs and note-taking apps (Obsidian, etc.). If you remain philosophically an IDE you are competing directly with VS Code (a juggernaut of a dev ecosystem, a huge plug-in marketplace, the best dev tools business in the world behind it), IntelliJ (excellent, full-featured IDEs with a huge enterprise installed base), etc. Notice the tools people are name-checking in this thread. VS Code "because they have to" but Sublime is what they want to replace. Which is exactly my point.

**Comment by GrabbenD** [comment]
> Just got bit by this and lost several files

Same here - 3 times in a week.

Some observations:
- XFS with Atomic Writes isn't enough to prevent the database from corrupting itself.
- Method for closing ZED is inconsequential as I've encountered this problem multiple times with either action:
  - Editor's Exit button
  - Computer's power button
  - Graceful kill _(default keybind to exit apps with tiling WM/Sway)_

## Workaround

You might be able to recover contents of unsaved files with:

```console
sqlite3 ~/.local/share/zed/db/0-stable/db.sqlite 'select contents from editors;'
```

---

Please keep us posted if there are any better way(s) to recover unsaved buffers 🙂

**Comment by Hermholtz** [comment]
We managed to gather 91 upvotes. Kindly please prioritize this. Thanks in advance!

**Comment by jonagh** [comment]
Agree. Being able to have "ephemeral" files is a very nice feature. Also being able to open random files from disk and have them automatically re-open next time is also very nice. And having to make sure a project is open (which may not have any relevance) is not intuitive.

Similar to the suggestion by @AldoMX - Zed could just have a "default" project. And considering that Zed's projects allow arbitrary/unrelated folders & files in them, I'm not sure if a project actually needs any folder association at all? (I don't know what's going on under the hood).

Zed could get rid of the no-project mode, and just use a "default" project automatically. It seems like this could be solved without a re-architecture.

**Comment by MRxRadex** [comment]
This is very important. Wanted to give it a try, but saw this behaviour and see no sense in using as a main tool. Using Zed as the text editor - can be an intro point to the product before using it in a workspace / project.

Same importance for restoring all opened files as saved and non saved.

**Comment by runxc1** [comment]
This is my primary use case for Zed to replace my text editor that I use for keeping track of temporary pieces of code I don't want in my primary driver (the real Visual Studio)

**Comment by babang1da** [comment]
please make it

**Comment by injust** [comment]
@echuber2 I think you're running into https://github.com/zed-industries/zed/issues/48468. You should see an error notification about failing to reopen the workspace. I believe that error is race-y and it will sometimes work. This affects non-project/worktree workspaces.

If you are working within a project/worktree, files should always reopen. If they don't, then you should open a new issue with reproduction steps.

**Comment by CreepyGnome** [comment]
I do not use workspaces not even sure what they are I open zed and just open files either as the default MD editor or by drag and drop there is not concept of workspaces or projects, it currently is just an adhoc editor with lots of tabs I want to keep track of and many I never want to save but do not want to lose any data from.   I was opening it would work out of the box by default like Notepad++ on windows or VS Code on any platform I guess.  If this requires a workspace or project concept to work "correctly" that could be a show stopper for me I don't want to have to set that up all the time as that is extra steps.

Also not even sure how to do workspace or project as there is no New/Save Workspace or New/Save Project option.  Seems like when opened it should just start in a default workspace maybe.  Make it impossible for Zed to operate outside of this workspace concept you mention.

**Comment by echuber2** [comment]
I think this was mostly fixed by https://github.com/zed-industries/zed/pull/54224 in the case of terminating the whole app, but closing an individual window with unsaved buffers might still instantly close and discard them all without warning.
